### PR TITLE
Enable reCaptcha after max failed attempt for Basic Authentication

### PR DESCRIPTION
**Pull request by bhagyasakalanka**
Related Issues: wso2/product-is#4992

### CODE REVIEW

File: components/org.wso2.carbon.identity.application.authenticator.basicauth/src/main/java/org/wso2/carbon/identity/application/authenticator/basicauth/BasicAuthenticator.java [hunk 1]
```diff
@@ -696,6 +700,30 @@ private String getCaptchaParams(String tenantDomain) {
...
                         }
+                    } else if (captchaEnabledConfigName.equals(connectorConfig.getName())) {
+                        if (Boolean.parseBoolean(connectorConfig.getValue())) {
+                            Property[] maxFailedConfig = BasicAuthenticatorDataHolder.getInstance()
+                                    .getIdentityGovernanceService()
+                                    .getConfiguration(new String[]{maxFailedAttemptCaptchaConfigName}, tenantDomain);
+                            Property maxFailedProperty = maxFailedConfig[0];
+                            try {
+                                if (Integer.valueOf(maxFailedProperty.getValue()) < failedLoginAttempts) {
+                                    captchaParams += BasicAuthenticatorConstants.RECAPTCHA_PARAM + "true";
```
Comment (sumedhe) [inline components/org.wso2.carbon.identity.application.authenticator.basicauth/src/main/java/org/wso2/carbon/identity/application/authenticator/basicauth/BasicAuthenticator.java:711]: Shall we add a debug log here too, when the no of failed attempts passed the max failed attempts

Comment (bhagyasakalanka) [reply components/org.wso2.carbon.identity.application.authenticator.basicauth/src/main/java/org/wso2/carbon/identity/application/authenticator/basicauth/BasicAuthenticator.java:715]: fixed fbecd0cf90e48bdbe48887baf899da44143a3f3d

**Review by ashensw — APPROVED** [review]

File: components/org.wso2.carbon.identity.application.authenticator.basicauth/src/main/java/org/wso2/carbon/identity/application/authenticator/basicauth/BasicAuthenticator.java [hunk 2]
```diff
@@ -696,6 +700,34 @@ private String getCaptchaParams(String tenantDomain) {
...
+                                } else {
+                                    if (log.isDebugEnabled()) {
+                                        log.debug("Number of failed attempts is less than max failed " +
+                                                "attempts for reCaptcha. Recaptcha will not be enforced");
+                                    }
+                                }
+                            } catch (NumberFormatException e) {
+                                log.error(String.format("Invalid Max failed attempts for reCaptcha: %s",
+                                        maxFailedProperty.getValue()), e);
```
Comment (somindatommy) [inline components/org.wso2.carbon.identity.application.authenticator.basicauth/src/main/java/org/wso2/carbon/identity/application/authenticator/basicauth/BasicAuthenticator.java:724]: What is the behavior if this error happens? Do we set a default value or end the flow?

File: components/org.wso2.carbon.identity.application.authenticator.basicauth/src/main/java/org/wso2/carbon/identity/application/authenticator/basicauth/BasicAuthenticator.java [hunk 3]
```diff
@@ -696,6 +700,34 @@ private String getCaptchaParams(String tenantDomain) {
...
+                                    if (log.isDebugEnabled()) {
+                                        log.debug("Number of failed attempts is higher than max failed " +
+                                                "attempts for reCaptcha. Recaptcha will be enforced");
+                                    }
+                                    captchaParams += BasicAuthenticatorConstants.RECAPTCHA_PARAM + "true";
+                                } else {
+                                    if (log.isDebugEnabled()) {
+                                        log.debug("Number of failed attempts is less than max failed " +
+                                                "attempts for reCaptcha. Recaptcha will not be enforced");
+                                    }
```
Comment (somindatommy) [inline components/org.wso2.carbon.identity.application.authenticator.basicauth/src/main/java/org/wso2/carbon/identity/application/authenticator/basicauth/BasicAuthenticator.java:720]: Can't we just end the flow if the attempts have not been exceeded?

File: components/org.wso2.carbon.identity.application.authenticator.basicauth/src/main/java/org/wso2/carbon/identity/application/authenticator/basicauth/BasicAuthenticator.java [hunk 4]
```diff
@@ -696,6 +700,34 @@ private String getCaptchaParams(String tenantDomain) {
...
+                                }
+                            } catch (NumberFormatException e) {
+                                log.error(String.format("Invalid Max failed attempts for reCaptcha: %s",
+                                        maxFailedProperty.getValue()), e);
+                            }
+                        } else {
+                            if (log.isDebugEnabled()) {
+                                log.debug("Enforcing recaptcha for SSO Login is not enabled.");
+                            }
```
Comment (somindatommy) [inline components/org.wso2.carbon.identity.application.authenticator.basicauth/src/main/java/org/wso2/carbon/identity/application/authenticator/basicauth/BasicAuthenticator.java:729]: Can't we just end the flow if not enabled?

File: components/org.wso2.carbon.identity.application.authenticator.basicauth/src/main/java/org/wso2/carbon/identity/application/authenticator/basicauth/BasicAuthenticator.java (hunk 3 above)
Comment (bhagyasakalanka) [reply components/org.wso2.carbon.identity.application.authenticator.basicauth/src/main/java/org/wso2/carbon/identity/application/authenticator/basicauth/BasicAuthenticator.java:720]: This is a login scenario. If the number of max failed attempts are higher than some value, we enforce a recaptcha and continue login process. If the max failed attempt is less than the value, we will not enforce the recaptcha and continue the login process. Hence we cannot end the flow in the else condition.

File: components/org.wso2.carbon.identity.application.authenticator.basicauth/src/main/java/org/wso2/carbon/identity/application/authenticator/basicauth/BasicAuthenticator.java (hunk 2 above)
Comment (bhagyasakalanka) [reply components/org.wso2.carbon.identity.application.authenticator.basicauth/src/main/java/org/wso2/carbon/identity/application/authenticator/basicauth/BasicAuthenticator.java:724]: Ideally maxAttepmtBeforeSuccess claim should be an integer. But from UI, we can set non integer values as well. Hence if a non-integer charactor is set,  this error would occur. If that is the case, then we continue login procedure without enforcing the recaptcha.

File: components/org.wso2.carbon.identity.application.authenticator.basicauth/src/main/java/org/wso2/carbon/identity/application/authenticator/basicauth/BasicAuthenticator.java (hunk 4 above)
Comment (bhagyasakalanka) [reply components/org.wso2.carbon.identity.application.authenticator.basicauth/src/main/java/org/wso2/carbon/identity/application/authenticator/basicauth/BasicAuthenticator.java:729]: Same case as in above comments

File: components/org.wso2.carbon.identity.application.authenticator.basicauth/src/main/java/org/wso2/carbon/identity/application/authenticator/basicauth/BasicAuthenticator.java (hunk 3 above)
Comment (somindatommy) [reply components/org.wso2.carbon.identity.application.authenticator.basicauth/src/main/java/org/wso2/carbon/identity/application/authenticator/basicauth/BasicAuthenticator.java:720]: What i meant was can't we return the methof from here?

File: components/org.wso2.carbon.identity.application.authenticator.basicauth/src/main/java/org/wso2/carbon/identity/application/authenticator/basicauth/BasicAuthenticator.java (hunk 2 above)
Comment (somindatommy) [reply components/org.wso2.carbon.identity.application.authenticator.basicauth/src/main/java/org/wso2/carbon/identity/application/authenticator/basicauth/BasicAuthenticator.java:724]: In that case don't we need to say that we are skipping the captcha flow due to this? The current error message says that it's not an int.

Comment (bhagyasakalanka) [reply components/org.wso2.carbon.identity.application.authenticator.basicauth/src/main/java/org/wso2/carbon/identity/application/authenticator/basicauth/BasicAuthenticator.java:724]: fixed https://github.com/wso2-extensions/identity-local-auth-basicauth/pull/118

File: components/org.wso2.carbon.identity.application.authenticator.basicauth/src/main/java/org/wso2/carbon/identity/application/authenticator/basicauth/BasicAuthenticator.java [hunk 5]
```diff
@@ -696,6 +700,34 @@ private String getCaptchaParams(String tenantDomain) {
...
+                                                "attempts for reCaptcha. Recaptcha will be enforced");
+                                    }
+                                    captchaParams += BasicAuthenticatorConstants.RECAPTCHA_PARAM + "true";
+                                } else {
+                                    if (log.isDebugEnabled()) {
+                                        log.debug("Number of failed attempts is less than max failed " +
+                                                "attempts for reCaptcha. Recaptcha will not be enforced");
+                                    }
+                                }
+                            } catch (NumberFormatException e) {
```
Comment (hasinthaindrajee) [inline components/org.wso2.carbon.identity.application.authenticator.basicauth/src/main/java/org/wso2/carbon/identity/application/authenticator/basicauth/BasicAuthenticator.java:722]: Is it safe to continue the flow and allow the user to access if this configuration is incorrect ?

Comment (bhagyasakalanka) [reply components/org.wso2.carbon.identity.application.authenticator.basicauth/src/main/java/org/wso2/carbon/identity/application/authenticator/basicauth/BasicAuthenticator.java:722]: yes there is an issue... This cause due to admin users error in configuring the value. So failing the login is not the best solution as I think. We can use the default value in such cases. Will fix this

Comment (ChamathNS) [reply components/org.wso2.carbon.identity.application.authenticator.basicauth/src/main/java/org/wso2/carbon/identity/application/authenticator/basicauth/BasicAuthenticator.java:722]: What's the value set for Max failed attempts if maxFailedProperty throws a NumberFormatException?

Comment (bhagyasakalanka) [reply components/org.wso2.carbon.identity.application.authenticator.basicauth/src/main/java/org/wso2/carbon/identity/application/authenticator/basicauth/BasicAuthenticator.java:722]: 3 is the default value

Comment (ChamathNS) [reply components/org.wso2.carbon.identity.application.authenticator.basicauth/src/main/java/org/wso2/carbon/identity/application/authenticator/basicauth/BasicAuthenticator.java:722]: If we are continuing with the default value if the user hasn't configured the value properly , let's mention that in the logs.

Comment (bhagyasakalanka) [reply components/org.wso2.carbon.identity.application.authenticator.basicauth/src/main/java/org/wso2/carbon/identity/application/authenticator/basicauth/BasicAuthenticator.java:722]: +1

Comment (bhagyasakalanka) [reply components/org.wso2.carbon.identity.application.authenticator.basicauth/src/main/java/org/wso2/carbon/identity/application/authenticator/basicauth/BasicAuthenticator.java:722]: fixed https://github.com/wso2-extensions/identity-local-auth-basicauth/pull/121

**Pull request merged**